rust/candid/src/bindings/motoko.rs

```diff
@@ -0,0 +1,236 @@
+            match t.as_ref() {
+                Service(ref serv) => doc.append(pp_service(serv)),
+                Var(ref s) => doc.append(s),
+                _ => unreachable!(),
+            }
+        }
+        Knot(_) | Unknown => unreachable!(),
+    }
+}
+
+fn pp_label(id: &Label) -> RcDoc {
+    match id {
+        Label::Named(str) => escape(str, false),
+        Label::Id(n) | Label::Unnamed(n) => str("_")
+            .append(RcDoc::as_string(n))
+            .append("_")
+            .append(RcDoc::space()),
+    }
+}
+
+fn pp_field(field: &Field) -> RcDoc {
+    pp_label(&field.id).append(" : ").append(pp_ty(&field.ty))
+}
+fn pp_variant(field: &Field) -> RcDoc {
+    let doc = str("#").append(pp_label(&field.id));
+    if field.ty != Type::Null {
+        doc.append(" : ").append(pp_ty(&field.ty))
+    } else {
+        doc
+    }
+}
+
+fn pp_function(func: &Function) -> RcDoc {
+    let args = pp_args(&func.args);
+    let rets = pp_args(&func.rets);
+    match func.modes.as_slice() {
+        [FuncMode::Oneway] => kwd("shared").append(args).append(" -> ").append("()"),
+        [FuncMode::Query] => kwd("shared query")
+            .append(args)
+            .append(" -> ")
+            .append("async ")
+            .append(rets),
+        [] => kwd("shared")
+            .append(args)
+            .append(" -> ")
+            .append("async ")
+            .append(rets),
+        _ => unreachable!(),
+    }
+    .nest(INDENT_SPACE)
+}
+fn pp_args(args: &[Type]) -> RcDoc {
+    match args {
+        [ty] => {
+            if is_tuple(ty) {
+                enclose("(", pp_ty(ty), ")")
+            } else {
+                pp_ty(ty)
+            }
+        }
+        _ => {
+            let doc = concat(args.iter().map(pp_ty), ",");
+            enclose("(", doc, ")")
+        }
+    }
+}
+
+fn pp_service(serv: &[(String, Type)]) -> RcDoc {
+    let doc = concat(
+        serv.iter()
+            .map(|(id, func)| escape(id, true).append(" : ").append(pp_ty(func))),
+        ";",
+    );
+    kwd("actor").append(enclose_space("{", doc, "}"))
+}
+
+fn pp_defs(env: &TypeEnv) -> RcDoc {
+    lines(env.0.iter().map(|(id, ty)| {
+        kwd("type")
+            .append(escape(id, false))
+            .append(" = ")
+            .append(pp_ty(ty))
+            .append(";")
+    }))
+}
+
+fn pp_actor(ty: &Type) -> RcDoc {
+    match ty {
+        Type::Service(ref serv) => pp_service(serv),
+        Type::Var(_) | Type::Class(_, _) => pp_ty(ty),
+        _ => unreachable!(),
+    }
+}
+
+pub fn compile(env: &TypeEnv, actor: &Option<Type>) -> String {
+    let header = r#"// This is a generated Motoko binding. Please use `import service "ic:canister_id"` instead to call canisters on the IC if possible.
+"#;
+    let doc = match actor {
+        None => pp_defs(env),
+        Some(actor) => {
+            let defs = pp_defs(env);
+            let actor = kwd("type _SERVICE =").append(pp_actor(actor));
```

Comment:
@nomeata Do you have a better idea to name the main actor? Put everything in a module or flat?

Reply:
I guess everything in a module makes sense, so that you can import the resulting `.mo` file, assuming that is the goal?

No good idea for a better name. When the optional name in
```
<actor> ::= service <id>? : (<tuptype> ->)? (<actortype> | <id>)
```
is present, one can use that. Otherwise, maybe `Self`, so that you can use `M.Self`? but `Service` isn’t bad either… no great idea, sorry.

Too bad we cannot somehow pun modules and types this, so that after
```
import M "generated_file.mo";
```
we can write `M` for the service type, and still get access to `M.Foo` for the types defined in `input.did`. It would also be closer to what `moc` does when you import a `.did` right now.

@crusso, do you have more ideas here?

Reply:
Yeah, `Self` is better, because it's sometimes a service, sometimes a service class.